src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ZERO_INDEX = 0;
+    private static final int USUALLY_SIZE = 1;
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
+    private void checkSize(int length) {
+        int capacity = elementData.length;
+        while (length + size > capacity) {
+            capacity = (int) (capacity * 1.5);
+            Object [] inheritElementData = new Object[capacity];
+            System.arraycopy(elementData, 0, inheritElementData, 0, size);
+            elementData = new Object[capacity];
+            elementData = inheritElementData;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < ZERO_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException("Undefined index");
+        }
+    }
+
+    private void checkAddIndex(int index) {
+        if (index > size || index < ZERO_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException("Undefined index");
+        }
+    }
 
     @Override
     public void add(T value) {
-
+        checkSize(USUALLY_SIZE);
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkAddIndex(index);
+        checkSize(USUALLY_SIZE);
+        if (index == size) {
+            elementData[size] = value;
+        } else if ((index < size) && (index >= ZERO_INDEX)) {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        checkSize(list.size());
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
```

Comment:
the size will be anyway checked in `add()` method
probably you don't need to call `checkSize(list.size()` here?